ICSharpCode.CodeConverter/CSharp/NodesVisitor.cs

```diff
@@ -1618,14 +1620,19 @@ public override CSharpSyntaxNode VisitBinaryExpression(VBSyntax.BinaryExpression
                     }
                 }
 
-                var lhs = (ExpressionSyntax)node.Left.Accept(TriviaConvertingVisitor);
-                var rhs = (ExpressionSyntax)node.Right.Accept(TriviaConvertingVisitor);
+                var lhs = AddExplicitConversion(node.Left, (ExpressionSyntax)node.Left.Accept(TriviaConvertingVisitor));
+                var rhs = AddExplicitConversion(node.Right, (ExpressionSyntax)node.Right.Accept(TriviaConvertingVisitor));
 
-                // e.g. VB DivideExpression "/" returns a double result for integer types (integer division is the "\" IntegerDivideExpression), so need to cast in C#
-                // see: https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/operators/floating-point-division-operator#remarks
-                if (node.IsKind(VBasic.SyntaxKind.DivideExpression) && node.Left.IsIntegralType(_semanticModel) && node.Right.IsIntegralType(_semanticModel)) {
-                    var doubleType = SyntaxFactory.PredefinedType(SyntaxFactory.Token(SyntaxKind.DoubleKeyword));
-                    rhs = SyntaxFactory.CastExpression(doubleType, rhs);
+                if (node.IsKind(VBasic.SyntaxKind.ConcatenateExpression)) {
+                    var stringType = _semanticModel.Compilation.GetTypeByMetadataName("System.String");
+                    var lhsTypeInfo = _semanticModel.GetTypeInfo(node.Left);
+                    var rhsTypeInfo = _semanticModel.GetTypeInfo(node.Right);
+                    if (lhsTypeInfo.Type.SpecialType != SpecialType.System_String &&
+                        lhsTypeInfo.ConvertedType.SpecialType != SpecialType.System_String &&
+                        rhsTypeInfo.Type.SpecialType != SpecialType.System_String &&
+                        rhsTypeInfo.ConvertedType.SpecialType != SpecialType.System_String) {
+                        lhs = AddExplicitConvertTo(node.Left, lhs, stringType);
+                    }
```

Comment:
It is done here instead so that eg.:

```vbnet
vBoolean & vObject
```

Becomes:

```C#
Conversions.ToString(vBoolean) + vObject
```

Rather than:

```C#
Conversions.ToString(vBoolean) + Conversions.ToString(vObject)
```

We could move this into `AddExplicitConvertTo()`  and have it convert both sides instead.

Reply:
Makes sense, let's leave it as is for now